pallets/settlement/src/lib.rs

```diff
@@ -1340,50 +1394,105 @@ impl<T: Trait> Module<T> {
                                     {
-                                        // Verify the proofs.
-                                        let mut rng = rng::Rng::default();
-                                        let result = TransactionValidator {}.verify_transaction(
-                                            &decoded_justified_tx,
-                                            &from_mercat,
-                                            &from_mercat_balance,
-                                            &to_mercat,
-                                            &to_mercat_balance,
-                                            &[],
-                                            &mut rng,
-                                        );
-                                        if let Ok((new_sender_balance, new_receiver_balance)) =
-                                            result
-                                        {
-                                            <ConfidentialAsset<T>>::update_mercat_account_balance(
-                                                &leg.from.did,
-                                                &leg.from_account_id,
-                                                new_sender_balance,
-                                            );
-                                            <ConfidentialAsset<T>>::update_mercat_account_balance(
-                                                &leg.to.did,
-                                                &leg.to_account_id,
-                                                new_receiver_balance,
-                                            );
+                                        // Get sender's current balance.
+                                        if let Ok((from_mercat, _)) = Self::get_account_and_balance(
+                                            &leg.from.did,
+                                            &leg.from_account_id,
+                                        ) {
+                                            // Verify the proofs.
+                                            let mut rng = rng::Rng::default();
+                                            let result = TransactionValidator {}
+                                                .verify_transaction(
+                                                    &decoded_justified_tx,
+                                                    &from_mercat,
+                                                    &from_mercat_pending_state,
+                                                    &to_mercat,
+                                                    &to_mercat_balance, // If the only reason we need this is updating balance, we should remove it.
+                                                    &[],
+                                                    &mut rng,
+                                                );
+
+                                            if let Ok((_, _)) = result {
```

Comment:
Assuming `result` is obtained with `map_err` and `?` as above, this can again be
```
let result = foo
    .map_err(|_| (leg_id, FailureReason::GetMercatAccount))?;
```
With a possibility of another error variant to differentiate between the two lines so that those don't emit the same error.